src/goto-instrument/dump_c.cpp

```diff
@@ -915,6 +945,66 @@ void dump_ct::cleanup_decl(
 
 /*******************************************************************\
 
+Function: dump_ct::collect_typedefs
+
+Inputs:
+
+Outputs:
+
+Purpose:
+
+\*******************************************************************/
+
+void dump_ct::collect_typedefs(const typet &type)
+{
+  if(type.id()==ID_code)
+  {
+    const code_typet &code_type=to_code_type(type);
+
+    collect_typedefs(code_type.return_type());
+    for(const auto &param : code_type.parameters())
+      collect_typedefs(param.type());
+  }
+  else if(type.id()==ID_pointer || type.id()==ID_array)
+  {
+    collect_typedefs(type.subtype());
+  }
+  else if(type.id()==ID_symbol)
+  {
+    const symbolt &symbol=
+      ns.lookup(to_symbol_type(type).get_identifier());
+    collect_typedefs(symbol.type);
+  }
+
+  const irep_idt &typedef_str=type.get(ID_C_typedef);
+
+  if(!typedef_str.empty())
+  {
+    std::pair<typedef_mapt::iterator, bool> entry=
+      typedef_map.insert({typedef_str, ""});
+
+    if(entry.second &&
+       (typedef_str=="__gnuc_va_list" ||
+        typedef_str == "va_list"))
+    {
+      system_headers.insert("stdarg.h");
+    }
+    else if(entry.second)
```

Comment:
This might be more clearly expressed as:

```C++
// If not already in the map
if(entry.second)
{
  if(typedef_str=="__gnuc_va_list" || typedef_str == "va_list")
  {
    system_headers.insert("stdarg.h");
  }
  else
  {
    typet t=type;
    t.remove(ID_C_typedef);
    std::ostringstream oss;
    oss << "typedef " << type_to_string(t) << " "
        << typedef_str << ';';
    entry.first->second=oss.str();
  }
}

```